ethereum/core/src/main/java/tech/pegasys/pantheon/ethereum/core/PendingTransactions.java

```diff
@@ -88,6 +88,18 @@ public PendingTransactions(
             "operation");
   }
 
+  List<Transaction> getLocalTransactions() {
```

Comment:
Maybe we could do it using parallel stream to optimize.
```suggestion
  List<Transaction> getLocalTransactions() {
    synchronized (pendingTransactions) {
      List<Transaction> localTransactions = new ArrayList<>();
      pendingTransactions
          .entrySet()
          .parallelStream()
          .filter(transaction -> transaction.getValue().isReceivedFromLocalSource())
          .forEach(transaction -> transaction.getValue().getTransaction());
      return localTransactions;
    }
  }
```

Reply:
Suggestion tip : cache the filtered list and store a boolean indicating whether the list is still consistent or not. Set it to false when a new transaction is received. We could have a method `invalidateCachedLocalTransactions()` to set this boolean to false when appropriate.

Reply:
Class level field to avoid instantiating `localTransactions` each time `localTransactions` is called ?

Reply:
I'd avoid any such caching until we know we have a performance issue.  We're currently having issues with memory usage so don't want to waste it on extra caches if they aren't necessary.